sql: TestTxnObeysTableModificationTime failed

[(sql).TestTxnObeysTableModificationTime failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1859247&tab=buildLog) on [release-20.1@7f397196e931c27494e2fccf54502921387c6de0](https://github.com/cockroachdb/cockroach/commits/7f397196e931c27494e2fccf54502921387c6de0):

```
I200407 18:30:18.262264 279845 sql/table.go:982  [n1,client=127.0.0.1:33930,hostssl,user=root] queued new schema change job for table 53, mutation 3
I200407 18:30:18.263318 279845 sql/sqlbase/structured.go:1554  [n1,client=127.0.0.1:33930,hostssl,user=root] publish: descID=53 (kv) version=8 mtime=1970-01-01 00:00:00 +0000 UTC
I200407 18:30:18.430406 279845 sql/event_log.go:132  [n1,client=127.0.0.1:33930,hostssl,user=root] Event: "drop_index", target: 53, info: {TableName:t.public.kv IndexName:foo Statement:DROP INDEX t.public.kv@foo User:root MutationID:3 CascadeDroppedViews:[]}
I200407 18:30:18.439579 279845 jobs/registry.go:285  [n1,client=127.0.0.1:33930,hostssl,user=root] scheduled jobs [544649439469993985]
I200407 18:30:18.538852 279867 kv/kvserver/replica_command.go:663  [n1,merge,s1,r33/1:/Table/53{-/2}] initiating a merge of r34:/{Table/53/2-Max} [(n1,s1):1, next=2, gen=2] into this range (lhs+rhs has (size=51 B+29 B=80 B qps=13.76+3.17=16.94qps) below threshold (size=128 MiB, qps=1250.00))
I200407 18:30:18.631513 279759 sql/lease.go:943  new lease: 53("kv") ver=8:1586284525.375693341,0, refcount=0
I200407 18:30:18.801863 277841 kv/kvserver/store_remove_replica.go:122  [n1,s1,r33/1:/Table/53{-/2}] removing replica r34/1
I200407 18:30:18.903095 277857 jobs/registry.go:1178  job 544649439469993985: resuming execution
I200407 18:30:18.931163 279952 jobs/registry.go:797  [n1] job 544649439469993985: stepping through state running with error <nil>
I200407 18:30:18.936277 279952 sql/schema_changer.go:442  [n1,job=544649439469993985,scExec] schema change on kv (53 v8) mutation 3 starting execution...
I200407 18:30:19.084069 279952 sql/sqlbase/structured.go:1554  [n1,job=544649439469993985,scExec] publish: descID=53 (kv) version=9 mtime=1970-01-01 00:00:00 +0000 UTC
I200407 18:30:19.272938 278027 server/status/runtime.go:498  [n1] runtime stats: 1.4 GiB RSS, 204 goroutines, 23 MiB/152 MiB/68 MiB GO alloc/idle/total, 31 MiB/81 MiB CGO alloc/total, 0.0 CGO/sec, 0.0/0.0 %(u/s)time, 0.0 %gc (948x), 554 KiB/554 KiB (r/w)net
I200407 18:30:19.273349 279952 sql/schema_changer.go:752  [n1,job=544649439469993985,scExec] waiting for a single version of table 53...
I200407 18:30:19.447600 279791 sql/lease.go:943  new lease: 53("kv") ver=9:1586284525.376693341,0, refcount=0
I200407 18:30:19.450114 279952 sql/lease.go:337  [n1,job=544649439469993985,scExec] waiting for 1 leases to expire: desc=[{kv 53 8}]
I200407 18:30:19.604507 279952 sql/schema_changer.go:754  [n1,job=544649439469993985,scExec] waiting for a single version of table 53... done (at v 9)
I200407 18:30:19.666004 279952 sql/backfill.go:171  [n1,job=544649439469993985,scExec] Running backfill for "kv", v=9, m=3
I200407 18:30:19.667166 279952 sql/backfill.go:287  [n1,job=544649439469993985,scExec] Completed backfill for "kv", v=9, m=3
I200407 18:30:19.958113 279952 sql/sqlbase/structured.go:1554  [n1,job=544649439469993985,scExec] publish: descID=53 (kv) version=10 mtime=1970-01-01 00:00:00 +0000 UTC
I200407 18:30:20.030294 280094 jobs/registry.go:797  [n1] job 544649444624302081: stepping through state running with error <nil>
I200407 18:30:20.127332 279952 sql/event_log.go:132  [n1,job=544649439469993985,scExec] Event: "finish_schema_change", target: 53, info: {MutationID:3}
I200407 18:30:20.128648 279952 sql/schema_changer.go:752  [n1,job=544649439469993985,scExec] waiting for a single version of table 53...
I200407 18:30:20.261057 280082 sql/lease.go:943  new lease: 53("kv") ver=10:1586284528.247515389,0, refcount=0
I200407 18:30:20.354742 279952 sql/lease.go:337  [n1,job=544649439469993985,scExec] waiting for 1 leases to expire: desc=[{kv 53 9}]
I200407 18:30:20.509089 279952 sql/schema_changer.go:754  [n1,job=544649439469993985,scExec] waiting for a single version of table 53... done (at v 10)
I200407 18:30:20.509674 279952 jobs/registry.go:797  [n1] job 544649439469993985: stepping through state succeeded with error <nil>
I200407 18:30:20.666059 279952 jobs/registry.go:952  [n1] job 544649439469993985: status succeeded after adoption finished
I200407 18:30:21.791540 279845 sql/event_log.go:132  [n1,client=127.0.0.1:33930,hostssl,user=root] Event: "set_cluster_setting", target: 0, info: {SettingName:kv.gc_ttl.strict_enforcement.enabled Value:DEFAULT User:root}
I200407 18:30:21.794488 277822 util/stop/stopper.go:539  quiescing
I200407 18:30:21.799272 278011 kv/kvserver/queue.go:578  [n1,s1] rate limited in gossip update (split): node unavailable; try another peer
W200407 18:30:21.800397 280105 kv/kvserver/intentresolver/intent_resolver.go:540  failed to cleanup transaction intents: failed to resolve intents: node unavailable; try another peer
I200407 18:30:21.801663 278011 kv/kvserver/queue.go:578  [n1,s1] rate limited in gossip update (merge): node unavailable; try another peer
I200407 18:30:21.898734 277944 kv/kvclient/kvcoord/transport_race.go:108  transport race promotion: ran 73 iterations on up to 1331 requests
I200407 18:30:22.414852 277872 kv/kvserver/queue.go:578  [n1,s1] rate limited in MaybeAdd (gc): node unavailable; try another peer
I200407 18:30:22.416003 277872 kv/kvserver/queue.go:578  [n1,s1] rate limited in MaybeAdd (replicate): node unavailable; try another peer
I200407 18:30:22.416326 277872 kv/kvserver/queue.go:578  [n1,s1] rate limited in MaybeAdd (replicaGC): node unavailable; try another peer
I200407 18:30:22.416602 277872 kv/kvserver/queue.go:578  [n1,s1] rate limited in MaybeAdd (raftlog): node unavailable; try another peer
I200407 18:30:22.416847 277872 kv/kvserver/queue.go:578  [n1,s1] rate limited in MaybeAdd (raftsnapshot): node unavailable; try another peer
I200407 18:30:22.417096 277872 kv/kvserver/queue.go:578  [n1,s1] rate limited in MaybeAdd (consistencyChecker): node unavailable; try another peer
I200407 18:30:22.417317 277872 kv/kvserver/queue.go:578  [n1,s1] rate limited in MaybeAdd (timeSeriesMaintenance): node unavailable; try another peer
W200407 18:30:22.534658 277855 jobs/registry.go:1215  job 544649444624302081: canceling due to liveness failure
E200407 18:30:22.535716 280094 jobs/registry.go:946  [n1] job 544649444624302081: adoption completed with error job 544649444624302081: node liveness error: restarting in background
W200407 18:30:22.536481 280094 kv/txn.go:603  [n1] failure aborting transaction: node unavailable; try another peer; abort caused by: context canceled
E200407 18:30:22.536957 280094 jobs/registry.go:950  [n1] job 544649444624302081: failed querying status: context canceled
--- FAIL: TestTxnObeysTableModificationTime (13.82s)
    lease_test.go:1084: pq: restart transaction: TransactionRetryWithProtoRefreshError: TransactionRetryError: retry txn (RETRY_COMMIT_DEADLINE_EXCEEDED - txn timestamp pushed too much; deadline exceeded by 91.120487ms (1586284218.199374240,2 > 1586284218.108253753,0)): "sql txn" meta={id=5a4994bf key=/Table/53/1/"c"/0 pri=0.00018529 epo=0 ts=1586284218.199374240,2 min=1586284218.065976370,0 seq=3} lock=true stat=PENDING rts=1586284218.199374240,2 wto=false max=1586284218.065976370,0


Failed to find issue assignee: 
old friend vivekmenezes is friendless
```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestTxnObeysTableModificationTime PKG=./pkg/sql TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestTxnObeysTableModificationTime.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql).TestTxnObeysTableModificationTime failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=1920248&tab=buildLog) on [release-20.1@27ed4fc1c3610bc57d49eaba9dfe209e4cd8ae83](https://github.com/cockroachdb/cockroach/commits/27ed4fc1c3610bc57d49eaba9dfe209e4cd8ae83):

```
I200504 21:55:24.815830 250846 kv/kvserver/store.go:2470  [n1,s1] 
** Compaction Stats [default] **
Level    Files   Size     Score Read(GB)  Rn(GB) Rnp1(GB) Write(GB) Wnew(GB) Moved(GB) W-Amp Rd(MB/s) Wr(MB/s) Comp(sec) CompMergeCPU(sec) Comp(cnt) Avg(sec) KeyIn KeyDrop
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Sum      0/0    0.00 KB   0.0      0.0     0.0      0.0       0.0      0.0       0.0   0.0      0.0      0.0      0.00              0.00         0    0.000       0      0
 Int      0/0    0.00 KB   0.0      0.0     0.0      0.0       0.0      0.0       0.0   0.0      0.0      0.0      0.00              0.00         0    0.000       0      0

** Compaction Stats [default] **
Priority    Files   Size     Score Read(GB)  Rn(GB) Rnp1(GB) Write(GB) Wnew(GB) Moved(GB) W-Amp Rd(MB/s) Wr(MB/s) Comp(sec) CompMergeCPU(sec) Comp(cnt) Avg(sec) KeyIn KeyDrop
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Uptime(secs): 21.8 total, 21.8 interval
Flush(GB): cumulative 0.000, interval 0.000
AddFile(GB): cumulative 0.000, interval 0.000
AddFile(Total Files): cumulative 0, interval 0
AddFile(L0 Files): cumulative 0, interval 0
AddFile(Keys): cumulative 0, interval 0
Cumulative compaction: 0.00 GB write, 0.00 MB/s write, 0.00 GB read, 0.00 MB/s read, 0.0 seconds
Interval compaction: 0.00 GB write, 0.00 MB/s write, 0.00 GB read, 0.00 MB/s read, 0.0 seconds
Stalls(count): 0 level0_slowdown, 0 level0_slowdown_with_compaction, 0 level0_numfiles, 0 level0_numfiles_with_compaction, 0 stop for pending_compaction_bytes, 0 slowdown for pending_compaction_bytes, 0 memtable_compaction, 0 memtable_slowdown, interval 0 total count
estimated_pending_compaction_bytes: 0 B
W200504 21:55:24.816320 252858 kv/kvserver/node_liveness.go:563  [n1,s1,r3/1:/System/{NodeLive…-tsd}] slow heartbeat took 5.7s
I200504 21:55:24.837141 252524 sql/event_log.go:132  [n1,job=552333816147410945,scExec] Event: "finish_schema_change", target: 53, info: {MutationID:3}
I200504 21:55:24.840181 252981 sql/lease.go:943  new lease: 53("kv") ver=10:1588629621.632125058,0, refcount=0
I200504 21:55:24.839769 250806 kv/kvserver/queue.go:578  [n1,s1,r23/1:/Table/2{7-8}] rate limited in MaybeAdd (merge): throttled on async limiting semaphore
I200504 21:55:24.854867 252524 sql/schema_changer.go:706  [n1,job=552333816147410945,scExec] waiting for a single version of table 53...
I200504 21:55:24.856782 252906 jobs/registry.go:806  [n1] job 552333816505991169: stepping through state running with error <nil>
I200504 21:55:24.857281 252524 sql/schema_changer.go:708  [n1,job=552333816147410945,scExec] waiting for a single version of table 53... done (at v 10)
I200504 21:55:24.858680 252524 jobs/registry.go:806  [n1] job 552333816147410945: stepping through state succeeded with error <nil>
I200504 21:55:24.863037 252524 jobs/registry.go:961  [n1] job 552333816147410945: status succeeded after adoption finished
I200504 21:55:25.949367 252020 sql/event_log.go:132  [n1,client=127.0.0.1:60658,hostssl,user=root] Event: "set_cluster_setting", target: 0, info: {SettingName:kv.gc_ttl.strict_enforcement.enabled Value:DEFAULT User:root}
I200504 21:55:25.951168 250637 util/stop/stopper.go:539  quiescing
W200504 21:55:25.952923 253071 kv/kvserver/intentresolver/intent_resolver.go:540  failed to cleanup transaction intents: failed to resolve intents: node unavailable; try another peer
I200504 21:55:26.147131 250752 kv/kvserver/queue.go:578  [n1,s1] rate limited in MaybeAdd (gc): node unavailable; try another peer
I200504 21:55:26.147208 250752 kv/kvserver/queue.go:578  [n1,s1] rate limited in MaybeAdd (split): node unavailable; try another peer
I200504 21:55:26.147231 250752 kv/kvserver/queue.go:578  [n1,s1] rate limited in MaybeAdd (replicate): node unavailable; try another peer
I200504 21:55:26.147243 250752 kv/kvserver/queue.go:578  [n1,s1] rate limited in MaybeAdd (replicaGC): node unavailable; try another peer
I200504 21:55:26.147253 250752 kv/kvserver/queue.go:578  [n1,s1] rate limited in MaybeAdd (raftlog): node unavailable; try another peer
I200504 21:55:26.147264 250752 kv/kvserver/queue.go:578  [n1,s1] rate limited in MaybeAdd (raftsnapshot): node unavailable; try another peer
I200504 21:55:26.147277 250752 kv/kvserver/queue.go:578  [n1,s1] rate limited in MaybeAdd (consistencyChecker): node unavailable; try another peer
I200504 21:55:26.147289 250752 kv/kvserver/queue.go:578  [n1,s1] rate limited in MaybeAdd (timeSeriesMaintenance): node unavailable; try another peer
W200504 21:55:26.157955 250684 jobs/registry.go:1224  job 552333816505991169: canceling due to liveness failure
E200504 21:55:26.158080 252906 jobs/registry.go:955  [n1] job 552333816505991169: adoption completed with error job 552333816505991169: node liveness error: restarting in background
W200504 21:55:26.158137 252906 kv/txn.go:607  [n1] failure aborting transaction: node unavailable; try another peer; abort caused by: context canceled
E200504 21:55:26.158148 252906 jobs/registry.go:959  [n1] job 552333816505991169: failed querying status: context canceled
--- FAIL: TestTxnObeysTableModificationTime (23.20s)
    lease_test.go:1084: pq: restart transaction: TransactionRetryWithProtoRefreshError: TransactionRetryError: retry txn (RETRY_COMMIT_DEADLINE_EXCEEDED - txn timestamp pushed too much; deadline exceeded by 18.556722539s (1588629322.404038378,2 > 1588629303.847315839,0)): "sql txn" meta={id=be9ebd7e key=/Table/53/1/"c"/0 pri=0.01551495 epo=0 ts=1588629322.404038378,2 min=1588629303.828277362,0 seq=3} lock=true stat=PENDING rts=1588629322.404038378,2 wto=false max=1588629303.828277362,0


Failed to find issue assignee: 
old friend vivekmenezes is friendless
```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestTxnObeysTableModificationTime PKG=./pkg/sql TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestTxnObeysTableModificationTime.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/catalog/lease/lease_test.go
+++ b/pkg/sql/catalog/lease/lease_test.go
@@ -39,6 +39,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/testcluster"
+	"github.com/cockroachdb/cockroach/pkg/util"
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
@@ -912,6 +913,19 @@ func TestTxnObeysTableModificationTime(t *testing.T) {
 	// Disable strict GC TTL enforcement because we're going to shove a zero-value
 	// TTL into the system with AddImmediateGCZoneConfig.
 	defer sqltestutils.DisableGCTTLStrictEnforcement(t, sqlDB)()
+
+	// This test intentionally relies on uncontended transactions not being pushed
+	// in order to verify what it claims to verify. The default closed timestamp
+	// interval in 20.1+ is 3s. When run under the race detector, the process can
+	// stall for upwards of 3s leading to the write transaction getting pushed.
+	//
+	// In order to mitigate that push, we increase the target_duration when the
+	// test is run under race.
+	if util.RaceEnabled {
+		_, err := sqlDB.Exec(
+			"SET CLUSTER SETTING kv.closed_timestamp.target_duration = '120s'")
+		require.NoError(t, err)
+	}
 
 	if _, err := sqlDB.Exec(`
 CREATE DATABASE t;

